docs/architecture/adr-083-libp2p-implementation.md

```diff
@@ -0,0 +1,163 @@
+
+## Alternative Approaches
+
+None. 
+
+The decision to adopt libp2p was part of ADR 73, and this ADR lays out
+an implementation plan and records a collection of decisions related
+to this implementation which have not been formally collected in any
+other form. 
+
+## Decisions
+
+### Feature Parity
+
+The goal of this project should be a p2p stack that has feature parity
+with the legacy stack. Everything that is possible with the p2p stack
+today should be possible in the future, eventually reducing the
+overhead of peer communication and complexity in the gossip/peer layer
+of the system. 
+
+In persuit of this goal, this project has two explicit *non-goals*:
+
+- Do not require wire compatibility with legacy stack.
+- Do not support dual-mode (legacy/libp2p) in a single node or network.
+
+There does not seem to be a compelling use case of having a release
+that contains both P2P stacks. While either of these goals could
+be theoretically possible, by eliminating both of these goals we can
+effectively reduce the scope of the project and increase our
+confidence.
+
+### Target Iterative Progress 
+
+Rather than produce an extensive design document (other than this
+record of decisions,) we will work iteratively in a topic
+branch--[`main/libp2p`](https://github.com/tendermint/tendermint/tree/main/libp2p)--to
+prototype the new libp2p integration. This work will aim to:
+
+- develop a working implementation (by way of the e2e tests) of the
+  p2p layer without requiring changes to the reactor implementations.
+- migrate the reactor tests after completing a prototype. 
+- address design questions as they come up, by amending this document
+  or in periodic conversations over patches. 
+- minimize or avoid any interface changes between the reactor
+  implementations and the peer layer. 
+
+### Use libp2p Idiomatically 
+
+We should endeavor to use Libp2p as idiomatically as possible and take
+advantage of features and developments available in the libp2p
+ecosystem.
+
+### Minimize P2P Code
+
+One of the largest benefits of this project is that we should be able
+to delete the vast majority of the existing p2p code, which should
+reduce the maintenance burden within Tendermint.
+
+### Defer Transport Decisions
+
+There are many high level aspects of p2p systems such as handshakes
+between nodes, transport protocols, which are fully configurable in
+the context of libp2p. There is no need to make decisions about these
+at this time, as these aspects of the system should be below the level
+of concern for tendermint.
+
+We will not make these decisions user-configurable without a specific
+need.  We will make it possible for users to pass a custom libp2p
+connection object (`host.Host`) with limited support, but will
+otherwise not expose configurations and customizations of different
+libp2p fundamentals.
+
+### Use GossipSub for Gossip Communication 
+
+An unfortunate consequence of the current p2p implementation is that
```

Comment:
I think this is debatable. For consensus, in particular, the gossip layer is very far from a "traditional" gossip layer. In fact, a more precise way to describe the consensus' "gossip" implementation is that peers kind of synchronize their consensus state object by exchanging messages. As a result (composition) of this peer-to-peer state synchronization we obtain something like gossip communication, not the opposite.

Reply:
I think that the properties of the system that you describe is really a side effect of the implementation rather than a design feature. As such, I think it's worth while to sort of take a step back from the current implementation and design, and avoid recapitulating earlier faults.